notes/pro/config2.md
Sure, here's the full Juniper configuration to redistribute as per the updated requirements:

```
interfaces {
    ge-0/0/0 {
        unit 0 {
            family inet {
                address 172.16.0.6/30;
            }
        }
    }
    xe-0/0/0 {
        unit 0 {
            family inet {
                address 192.168.1.1/24;
            }
        }
    }
    xe-0/1/3 {
        unit 0 {
            family inet {
                address 172.16.0.2/30;
            }
        }
    }
}

routing-options {
    router-id 10.0.0.1;
}

protocols {
    ospf {
        area 0.0.0.0 {
            interface xe-0/0/0.0 {
                passive;
                metric 10;
            }
            interface xe-0/1/3.0 {
                passive;
                metric 10;
            }
        }
    }

    bgp {
        group ebgp {
            type external;
            multihop {
                ttl 3;
            }
            export export-to-ospf;
            neighbor 172.16.0.5 {
                remote-as 2;
            }
            neighbor 172.16.0.1 {
                remote-as 3;
            }
        }
        group ibgp {
            type internal;
            export export-to-ospf;
            neighbor 10.0.0.2 {
                peer-as 1;
            }
            neighbor 10.0.0.3 {
                peer-as 1;
            }
        }
        group rr {
            type internal;
            cluster 1;
            neighbor 10.0.0.2 {
                remote-as 1;
                route-reflector-client;
            }
            neighbor 10.0.0.3 {
                remote-as 1;
                route-reflector-client;
            }
        }
    }
}

policy-options {
    policy-statement export-to-ospf {
        from protocol bgp;
        then {
            as-path-prepend "2 ";
            route-filter 0.0.0.0/0 prefix-length-range /0-/32;
            propagate-nexthop;
            accept;
        }
    }
}
``` 



notes/pro/set/config.md
jUNIPER MX204 routers

# Set hostname and enable SSH
set system host-name router-1
set system services ssh root-login allow
set system services ssh protocol-version v2

set system host-name router-2
set system services ssh root-login allow
set system services ssh protocol-version v2

set system host-name router-3
set system services ssh root-login allow
set system services ssh protocol-version v2

# Configure interfaces on Router-1
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.6/30

# Configure interfaces on Router-2
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.5/30
set interfaces ge-0/0/0 unit 0 family inet address 192.168.1.254/24
set interfaces xe-0/0/0 unit 0 family inet address 172.16.0.1/30

# Configure interfaces on Router-3
set interfaces xe-0/1/3 unit 0 family inet address 172.16.0.2/30

# Configure OSPF on Router-1
set protocols ospf area 0.0.0.0 interface ge-0/0/0.0
set protocols ospf area 0.0.0.0 interface lo0.0

# Configure OSPF on Router-2
set protocols ospf area 0.0.0.0 interface ge-0/0/0.0
set protocols ospf area 0.0.0.0 interface xe-0/0/0.0
set protocols ospf area 0.0.0.0 interface ge-0/0/0.1
set protocols ospf area 0.0.0.0 interface lo0.0

# Configure OSPF on Router-3
set protocols ospf area 0.0.0.0 interface xe-0/1/3.0
set protocols ospf area 0.0.0.0 interface lo0.0

# Configure BGP on Router-2
set protocols bgp group rr type internal
set protocols bgp group rr local-address 192.168.1.254
set protocols bgp group rr family inet unicast
set protocols bgp group rr cluster 1
set protocols bgp group rr neighbor 172.16.0.5
set protocols bgp group rr neighbor 172.16.0.2

# Configure BGP on Router-1 and Router-3
set protocols bgp group ext type external
set protocols bgp group ext family inet unicast
set protocols bgp group ext multihop
set protocols bgp group ext peer-as 65003
set protocols bgp group ext local-address 172.16.0.6
set protocols bgp group ext neighbor 172.16.0.5
set protocols bgp group ext neighbor 172.16.0.2

# Configure BGP on Route-2 (as a route reflector)
set protocols bgp group client type internal
set protocols bgp group client local-address 192.168.1.254
set protocols bgp group client family inet unicast
set protocols bgp group client cluster 1
set protocols bgp group client cluster-type internal
set protocols bgp group client cluster-id 1.1.1.1
set protocols bgp group client neighbor 172.16.0.6
set protocols bgp group client neighbor 172

###############
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.5/30
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.5/30
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.5/30
set interfaces xe-0/0/0 unit 0 family inet address 172.16.0.1/30
set interfaces xe-0/1/3 unit 0 family inet address 172.16.0.2/30

set protocols bgp group ebgp type external
set protocols bgp group ebgp export export-ebgp
set protocols bgp group ebgp peer-as 65001
set protocols bgp group ebgp neighbor 172.16.0.5 multihop ttl 255
set protocols bgp group ebgp neighbor 172.16.0.5 hold-time 30
set protocols bgp group ebgp neighbor 172.16.0.5 authentication md5 mypassword
set protocols bgp group ebgp neighbor 172.16.0.5 local-address 172.16.0.6

set protocols bgp group ibgp type internal
set protocols bgp group ibgp export export-ibgp
set protocols bgp group ibgp peer-as 65000
set protocols bgp group ibgp neighbor 192.168.0.1 peer-as 65001
set protocols bgp group ibgp neighbor 192.168.0.1 local-address 192.168.0.2
set protocols bgp group ibgp neighbor 192.168.0.1 import policy ibgp-in

set policy-options policy-statement export-ebgp term 1 from protocol direct
set policy-options policy-statement export-ebgp term 1 then accept

set policy-options policy-statement export-ibgp term 1 from protocol direct
set policy-options policy-statement export-ibgp term 1 then accept

set protocols ospf area 0.0.0.0 interface lo0.0
set protocols ospf area 0.0.0.0 interface ge-0/0/0
set protocols ospf area 0.0.0.0 interface xe-0/1/3

set routing-options static route 0.0.0.0/0 next-hop 192.168.1.1


############

# Set interfaces for router-1
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.6/30

# Set interfaces for router-r2
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.5/30
set interfaces ge-0/0/0 unit 0 family inet address 192.168.1.254/24
set interfaces xe-0/0/0/35 unit 0 family inet address 172.16.0.1/30

# Set interfaces for router-r3
set interfaces xe-0/1/3 unit 0 family inet address 172.16.0.2/30

# Set OSPF for router-1
set protocols ospf area 0.0.0.0 interface ge-0/0/0.0
set protocols ospf area 0.0.0.0 interface lo0.0 passive

# Set OSPF for router-r2
set protocols ospf area 0.0.0.0 interface ge-0/0/0.0
set protocols ospf area 0.0.0.0 interface ge-0/0/0.1
set protocols ospf area 0.0.0.0 interface xe-0/0/0/35.0

# Set OSPF for router-r3
set protocols ospf area 0.0.0.0 interface xe-0/1/3.0

# Set BGP for router-r2
set protocols bgp group EBGP type external
set protocols bgp group EBGP peer-as 65002
set protocols bgp group EBGP neighbor 172.16.0.6 multihop ttl 2
set protocols bgp group EBGP neighbor 172.16.0.6 export export-to-ospf
set protocols bgp group EBGP neighbor 172.16.0.6 import import-from-ospf
set protocols bgp group EBGP neighbor 172.16.0.6 description "peer to router-1"

# Set BGP for router-r2 as a route reflector
set protocols bgp group iBGP type internal
set protocols bgp group iBGP local-address 192.168.1.254
set protocols bgp group iBGP cluster 1
set protocols bgp group iBGP neighbor 172.16.0.2 peer-as 65003
set protocols bgp group iBGP neighbor 172.16.0.2 route-reflector-client
set protocols bgp group iBGP neighbor 172.16.0.6 peer-as 65001
set protocols bgp group iBGP neighbor 172.16.0.6 route-reflector-client

# Set policy options
set policy-options policy-statement export-to-ospf term term1 from protocol bgp
set policy-options policy-statement export-to-ospf term term1 then accept
set policy-options policy-statement import-from-ospf term term1 from protocol ospf
set policy-options policy-statement import-from-ospf term term1 then accept

# Set routing options
set routing-options static route 0.0.0.0/0 next-hop 192.168.1.1



R1#
ping 

notes/pro/set/final/config.sh
# IP ROUTING

For Router-1:
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.6/30
set routing-options static route 192.168.1.0/24 next-hop 172.16.0.5
set routing-options static route 192.207.172.51/32 next-hop 172.16.0.5

For Router-2(B):
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.5/30
set interfaces xe-0/0/0/35 unit 0 family inet address 172.16.0.1/30
set interfaces xe-0/1/3 unit 0 family inet address 172.16.0.2/30
set interfaces ge-0/0/1 unit 0 family inet address 192.168.1.254/24
set routing-options static route 0.0.0.0/0 next-hop 192.168.1.1

For Router-r3:
set interfaces xe-0/1/3 unit 0 family inet address 172.16.0.3/30
set routing-options static route 172.16.0.0/30 next-hop 172.16.0.2
set routing-options static route 192.207.172.51/32 next-hop 172.16.0.2


## OSPF AND BGP 
router-r1
set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.6/30

set interfaces ge-0/0/0 unit 0 family inet address 172.16.0.5/30
set interfaces xe-0/0/0/35 unit 0 family inet address 172.16.0.1/30

R3
set interfaces xe-0/1/3 unit 0 family inet address 172.16.0.2/30

# set interfaces ge-0/0/0 unit 0 family inet address 192.168.1.254/24

set routing-options router-id 1.1.1.1


set protocols bgp group ebgp type external
set protocols bgp group ebgp peer-as 2
set protocols bgp group ebgp neighbor 172.16.0.5
set protocols bgp group ebgp family inet unicast

## Router r2
set protocols bgp group ibgp type internal
set protocols bgp group ibgp local-address 192.168.1.254
set protocols bgp group ibgp family inet unicast
set protocols bgp group ibgp export ibgp-to-ospf
set protocols bgp group ibgp neighbor 172.16.0.6
# set protocols bgp group ibgp neighbor 172.16.0.2

set protocols ospf area 0.0.0.0 interface ge-0/0/0.0 interface-type p2p
set protocols ospf area 0.0.0.0 interface xe-0/0/0/35.0 interface-type p2p
set protocols ospf area 0.0.0.0 interface xe-0/1/3.0 interface-type p2p

set policy-options policy-statement bgp-to-ospf term bgp-to-ospf from protocol bgp
set policy-options policy-statement bgp-to-ospf term bgp-to-ospf from community bgp-to-ospf
set policy-options policy-statement bgp-to-ospf term bgp-to-ospf then accept
set policy-options policy-statement bgp-to-ospf term default then reject

set policy-options policy-statement ospf-to-bgp term ospf-to-bgp from protocol ospf
set policy-options policy-statement ospf-to-bgp term ospf-to-bgp then community add ospf-to-bgp
set policy-options policy-statement ospf-to-bgp term ospf-to-bgp then accept

set policy-options policy-statement loop-filter term reject-from-bgp from protocol bgp
set policy-options policy-statement loop-filter term reject-from-bgp from community bgp-to-ospf
set policy-options policy-statement loop-filter term reject-from-bgp then reject

set policy-options policy-statement loop-filter term accept-all then accept

set protocols ospf export ospf-to-bgp
set protocols ospf import loop-filter
set protocols bgp group ibgp import bgp-to-ospf
set protocols bgp group ibgp export ospf-to-bgp




# Set OSPF for router-1
set protocols ospf area 0.0.0.0 interface ge-0/0/0.0
set protocols ospf area 0.0.0.0 interface lo0.0 passive

# Set OSPF for router-r2
set protocols ospf area 0.0.0.0 interface ge-0/0/0.0
# set protocols ospf area 0.0.0.0 interface ge-0/0/0.1
# set protocols ospf area 0.0.0.0 interface xe-0/0/0/35.

# Set OSPF for router-r3 - no ospf
# set protocols ospf area 0.0.0.0 interface xe-0/1/3.0


# optimise ospf
set protocols ospf interface ge-0/0/0 hello-interval 2
set protocols ospf interface ge-0/0/0 dead-interval 8

# router r2

